addons/account/models/account_move.py

```diff
@@ -2561,20 +2561,51 @@ def _is_readonly(self):
     # -------------------------------------------------------------------------
     def _is_eligible_for_early_payment_discount(self, currency, reference_date):
         self.ensure_one()
-        payment_terms = self.line_ids.filtered(lambda line: line.display_type == 'payment_term')
-        return self.currency_id == currency \
-            and self.move_type in self._early_payment_discount_move_types() \
-            and self.invoice_payment_term_id.early_discount \
-            and (
-                not reference_date
-                or not self.invoice_date
-                or (
-                    (existing_discount_date := fields.first(payment_terms).discount_date)
-                    and
-                    reference_date <= existing_discount_date
-                )
-            ) \
-            and not (payment_terms.sudo().matched_debit_ids + payment_terms.sudo().matched_credit_ids)
+        return self._is_eligible_for_early_payment_discount_batched(currency, reference_date)[self.id]
+
+    def _is_eligible_for_early_payment_discount_batched(self, currency, reference_date):
+        is_eligible_by_move_id = {move.id: False for move in self}
+        move_types_for_epd = self._early_payment_discount_move_types()
+        eligible_moves = self.filtered(lambda m:
+                                       m.currency_id == currency
+                                       and m.move_type in move_types_for_epd
+                                       and m.invoice_payment_term_id.early_discount
+                                       )
+        if not eligible_moves:
+            return is_eligible_by_move_id
+
+        payment_term_lines = self.env['account.move.line'].search([
```

Comment:
Does a search really improve the perfs?
Is this what is causing the memory error, because we read too many lines?
Aren't those lines read anyway?